controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +108,308 @@ func NewReconciler(
-		if err := patchHelper.Patch(ctx, obj); err != nil {
+		if err := patchHelper.Patch(ctx, vmGroupPublishRequest); err != nil {
 			if reterr == nil {
 				reterr = err
 			}
-			objCtx.Logger.Error(err, "patch failed")
+			vmGroupPublishRequestCtx.Logger.Error(err, "patch failed")
 		}
 	}()
 
-	if !obj.DeletionTimestamp.IsZero() {
-		return r.ReconcileDelete(objCtx)
+	if !vmGroupPublishRequest.DeletionTimestamp.IsZero() {
+		return r.ReconcileDelete(vmGroupPublishRequestCtx)
 	}
 
-	return r.ReconcileNormal(objCtx)
+	return pkgerr.ResultFromError(r.ReconcileNormal(vmGroupPublishRequestCtx))
 }
 
 func (r *Reconciler) ReconcileDelete(
 	ctx *pkgctx.VirtualMachineGroupPublishRequestContext) (ctrl.Result, error) {
-
 	ctx.Logger.Info("Reconciling VirtualMachineGroupPublishRequest Deletion")
 
-	if controllerutil.ContainsFinalizer(ctx.Obj, finalizerName) {
-		controllerutil.RemoveFinalizer(ctx.Obj, finalizerName)
+	if controllerutil.ContainsFinalizer(ctx.VMGroupPublishRequest, finalizerName) {
+		ctx.Logger.Info("Removing owned VirtualMachinePublishRequests")
+		if err := r.deleteVMPublishRequests(ctx); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		controllerutil.RemoveFinalizer(ctx.VMGroupPublishRequest, finalizerName)
 		ctx.Logger.Info("Removed finalizer for VirtualMachineGroupPublishRequest")
 	}
 	ctx.Logger.V(4).Info("Finished Reconciling VirtualMachineGroupPublishRequest Deletion")
 	return ctrl.Result{}, nil
 }
 
-func (r *Reconciler) ReconcileNormal(
-	ctx *pkgctx.VirtualMachineGroupPublishRequestContext) (_ ctrl.Result, reterr error) {
+func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineGroupPublishRequestContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineGroupPublishRequest")
+	vmGroupPublishReq := ctx.VMGroupPublishRequest
 
-	ctx.Logger.Info("Reconciling VirtualMachinePublishRequest")
-	vmGroupPublishReq := ctx.Obj
+	if controllerutil.AddFinalizer(vmGroupPublishReq, finalizerName) {
+		ctx.Logger.Info("Added finalizer for VirtualMachineGroupPublishRequest")
+		return nil
+	}
 
-	if !controllerutil.ContainsFinalizer(vmGroupPublishReq, finalizerName) {
-		controllerutil.AddFinalizer(vmGroupPublishReq, finalizerName)
-		return ctrl.Result{}, reterr
+	if conditions.IsTrue(vmGroupPublishReq, vmopv1.VirtualMachineGroupPublishRequestConditionComplete) {
+		return r.reconcileSpecTTL(ctx)
 	}
 
-	return ctrl.Result{}, nil
+	if err := r.verifySpec(vmGroupPublishReq); err != nil {
+		return err
+	}
+
+	return r.reconcileStatus(ctx)
+}
+
+func (r *Reconciler) reconcileStatus(ctx *pkgctx.VirtualMachineGroupPublishRequestContext) error {
+	if ctx.VMGroupPublishRequest.Status.StartTime.IsZero() {
+		ctx.VMGroupPublishRequest.Status.StartTime = metav1.NewTime(time.Now())
+	}
+
+	existReqsMap, pendingReqsSet, completedReqsSet, err := r.getVMPublishRequests(ctx)
+	if err != nil {
+		return err
+	}
+
+	r.reconcileStatusImages(ctx, existReqsMap)
+
+	desiredReqsSet := sets.New(ctx.VMGroupPublishRequest.Spec.VirtualMachines...)
```

Comment:
The API docs mentioned that "If omitted, this field defaults to the names of all of the VMs currently a member of the group, either directly or indirectly via a nested group." Will there be a follow-up change to address this?

Reply:
Yeah, this alone will spec.target and spec.source will be done in the webhook changes. We have VMSVC-2895 for tracking it.